packages/typegpu/src/tgpuBindGroupLayout.ts

```diff
@@ -0,0 +1,449 @@
+    TgpuLayoutEntry | null
+  >,
+> implements TgpuBindGroup<Entries>
+{
+  public readonly resourceType = 'bind-group' as const;
+
+  constructor(
+    public readonly layout: TgpuBindGroupLayout<Entries>,
+    public readonly entries: {
+      [K in keyof Entries]: LayoutEntryToInput<Entries[K]>;
+    },
+  ) {
+    // Checking if all entries are present.
+    for (const key of Object.keys(layout.entries)) {
+      if (layout.entries[key] !== null && !(key in entries)) {
+        throw new MissingBindingError(layout.label, key);
+      }
+    }
+  }
+
+  public unwrap(unwrapper: Unwrapper): GPUBindGroup {
+    const unwrapped = unwrapper.device.createBindGroup({
+      label: this.layout.label ?? '',
+      layout: unwrapper.unwrap(this.layout),
+      entries: Object.entries(this.layout.entries)
+        .map(([key, entry], idx) => {
+          if (entry === null) {
+            return null;
+          }
+
+          const value = this.entries[key];
+
+          if (value === undefined) {
+            throw new Error(
+              `'${key}' is a resource required to populate bind group layout '${this.layout.label ?? '<unnamed>'}'.`,
+            );
+          }
+
+          if ('uniform' in entry) {
+            let resource: GPUBufferBinding;
+
+            if (isBuffer(value)) {
+              if (!isUsableAsUniform(value)) {
+                throw new NotUniformError(value);
+              }
+              resource = { buffer: unwrapper.unwrap(value) };
+            } else if (isBufferUsage(value)) {
+              if (!isUsableAsUniform(value.allocatable)) {
+                throw new NotUniformError(value.allocatable);
+              }
+              resource = { buffer: unwrapper.unwrap(value.allocatable) };
+            } else {
+              resource = { buffer: value as GPUBuffer };
+            }
+
+            return {
+              binding: idx,
+              resource,
+            };
+          }
+
+          if ('storage' in entry) {
+            let resource: GPUBufferBinding;
+
+            if (isBuffer(value)) {
+              if (!isUsableAsStorage(value)) {
+                throw new NotUniformError(value);
+              }
+              resource = { buffer: unwrapper.unwrap(value) };
+            } else if (isBufferUsage(value)) {
+              if (!isUsableAsStorage(value.allocatable)) {
+                throw new NotUniformError(value.allocatable);
+              }
+              resource = { buffer: unwrapper.unwrap(value.allocatable) };
+            } else {
+              resource = { buffer: value as GPUBuffer };
+            }
+
+            return {
+              binding: idx,
+              resource,
+            };
+          }
+
+          if (
+            'texture' in entry ||
+            'storageTexture' in entry ||
+            'sampler' in entry ||
+            'externalTexture' in entry
+          ) {
+            return {
+              binding: idx,
+              resource: value as GPUSampler,
```

Comment:
`as GPUSampler` to make TS not complain, but the type is inferred correctly regardless?

and so far we only support raw WebGPU samplers and textures here, not our tgpu objects here, right?

Reply:
Yes, we only support raw samplers and textures for now, until we are sure our tgpu equivalents also match 1-to-1 with WebGPU.